chore(suite-native): biometrics reducer

juriczech opened the pull request:
## Description

- Migrates Suite Native biometrics state from Jotai atoms to Redux so biometrics is managed consistently with the rest of the app.
- Adds a persisted `biometrics` Redux slice with selectors, thunks, and MMKV persistence for `isBiometricsEnabled`.
- Adds a migration from the legacy `isBiometricsOptionEnabled` Jotai storage value into Redux persistence.
- Reworks authentication, settings toggling, and app state handling through Redux thunks.
- Updates onboarding, settings, app freeze handling, and analytics to use Redux selectors.
- Removes obsolete Jotai biometrics hooks/storage and updates package dependencies.
- Big win is that now we can cover the logic with unit tests 🎉 

<!-- PREVIEW-LINKS:SECTION:START -->
## 🌐 Preview deployments

<!-- PREVIEW-LINK:SUITE-WEB:START -->
🌐 **Suite Web preview:** [https://dev.suite.sldev.cz/suite-web/chore/native/biometrics-redux/web/](https://dev.suite.sldev.cz/suite-web/chore/native/biometrics-redux/web/)
<!-- PREVIEW-LINK:SUITE-WEB:END -->
<!-- PREVIEW-LINKS:SECTION:END -->

<!-- CURRENTS-LINKS:SECTION:START -->
## 🔍 Currents Test Results

<!-- CURRENTS-LINK:DESKTOP:START -->
🔍 **Suite desktop test results:** [View in Currents](https://app.currents.dev/projects/4ytF0E/runs?branch=chore/native/biometrics-redux&lookback=7)
<!-- CURRENTS-LINK:DESKTOP:END -->
<!-- CURRENTS-LINK:WEB:START -->
🔍 **Suite web test results:** [View in Currents](https://app.currents.dev/projects/Og0NOQ/runs?branch=chore/native/biometrics-redux&lookback=7)
<!-- CURRENTS-LINK:WEB:END -->
<!-- CURRENTS-LINKS:SECTION:END -->

<!-- QUARANTINE-LIST:HEADING:START -->
## 🔒 Quarantined E2E Tests
<!-- QUARANTINE-LIST:HEADING:END -->

<!-- QUARANTINE-LIST:web:START -->
<details><summary>Trezor Suite (web) — 6 test(s)</summary>

| Test | Type |
|------|------|
| Account types suite > Add-account-types-non-BTC-coins | 🤖 auto |
| Account metadata > dropbox provider | 🤖 auto |
| Quarantine test: "Onboarding - create wallet,Success (basic)" | 🙋 manual |
| Quarantine test: "Database migration,Db migration between: release/22.5/web => develop/web" | 🙋 manual |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |
| Quarantine test: "Recovery T2T1 - dry run,Recovery after partial recovery" | 🙋 manual |

</details>

_Updated: 2026-06-03T12:33:57.168Z • 6 test(s) total_
<!-- QUARANTINE-LIST:web:END -->

<!-- QUARANTINE-LIST:desktop:START -->
<details><summary>Trezor Suite (desktop) — 4 test(s)</summary>

| Test | Type |
|------|------|
| Labeling migration > Migration from local file | 🤖 auto |
| Account types suite > Add-account-types-non-BTC-coins | 🤖 auto |
| Onboarding - create wallet > Success (basic) | 🤖 auto |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |

</details>

_Updated: 2026-06-03T12:30:41.736Z • 4 test(s) total_
<!-- QUARANTINE-LIST:desktop:END -->

juriczech committed:
chore(suite-native): biometrics reducer

juriczech committed:
chore(suite-native): unify getIsFeatureAvailable

juriczech committed:
chore(suite-native): toggle biometrics settings

juriczech committed:
feat(suite-native): trigger manual authentication

juriczech committed:
chore(suite-native): simplify useBiometrics and doAuthentication

juriczech committed:
feat(suite-native): remove isBiometricsOverlayVisible

juriczech committed:
chore(suite-native): isAuthenticating, tweak effect

juriczech committed:
feat(suite-native): handleBiometricsAppStateChangeThunk

juriczech committed:
feat(suite-native): toggle settings tweaks

juriczech committed:
chore(suite-native): migration

juriczech committed:
fix(suite-native): biometrics unavailable alert

juriczech committed:
fix(suite-native): biometrics selector

matejkriz commented on suite-native/storage/src/tests/migrations/biometricsV1.test.ts:
wrong filename V2 vs V1

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/commits/85b1d1bb602ace133d06f147a2ae912d797d5bc4

matejkriz commented on suite-native/storage/src/migrations/biometrics/v1.ts:
```diff
@@ -0,0 +1,45 @@
+export const migrateBiometricsAtomToRedux = (oldState: unknown) => {
+    if (!oldState || !isPersistedState(oldState)) {
+        return oldState as BiometricsPersistedState;
+    }
+
+    const persistedState = oldState as BiometricsPersistedState;
+
+    const storedValue = parseStoredBoolean(
+        unecryptedJotaiStorage.getString(BIOMETRICS_PERSISTED_ATOM_STORAGE_KEY),
+    );
+
+    if (typeof storedValue !== 'boolean') {
+        return persistedState;
+    }
+
+    unecryptedJotaiStorage.remove(BIOMETRICS_PERSISTED_ATOM_STORAGE_KEY);
+
+    return {
+        ...persistedState,
+        isBiometricsEnabled: storedValue,
+    };
```
In the initial migration it's better to return `undefined` instead of undefined oldState, than you don't need `as` at all:
```
export const migrateBiometricsAtomToRedux = (oldState: unknown) => {
    if (!oldState || !isPersistedState(oldState)) {
        return undefined;
    }

    const storedValue = parseStoredBoolean(
        unecryptedJotaiStorage.getString(BIOMETRICS_PERSISTED_ATOM_STORAGE_KEY),
    );

    if (typeof storedValue !== 'boolean') {
        return undefined;
    }

    unecryptedJotaiStorage.remove(BIOMETRICS_PERSISTED_ATOM_STORAGE_KEY);

    return {
        ...oldState,
        isBiometricsEnabled: storedValue,
    };
};
```

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/commits/85b1d1bb602ace133d06f147a2ae912d797d5bc4

Lemonexe replied:
Still relevant.

juriczech committed:
fix(suite-native): prompt biometrics on cold start

Previously the 'active' branch in handleBiometricsAppStateChangeThunk set
isUserAuthenticated=true whenever shouldRevokeAuth was falsy. On cold
start goneToBackgroundAtTimestamp is null, which made shouldRevokeAuth
falsy, so the user was silently marked authenticated without ever being
prompted.

Now restore prior auth only when returning from a real backgrounding
within the keep-logged-in window. Otherwise fall through to
authenticateUserThunk so the prompt actually fires.

juriczech committed:
refactor(suite-native): use canonical biometrics selectors in slice

The slice was redeclaring BiometricsSliceRootState plus five selectors
that already live in biometricsSelectors.ts. Drop the duplicates and
import from the selectors file so there is one source of truth.

juriczech committed:
chore(suite-native): drop dead blur/focus scaffolding

The blurSubscription / focusSubscription blocks were only console.logs
behind comments and never wired up. Remove them.

juriczech committed:
test(suite-native): expand biometrics v1 migration coverage

Add cases for the actually-interesting true value, malformed JSON,
non-boolean parsed values, and oldState that does not match the
PersistedState shape. Annotate the migration return type explicitly.

juriczech committed:
chore(suite-native): align biometrics dependencies with usage

Drop jotai, services, and storage dependencies (unused since the Redux
migration), declare the previously implicit @suite-common/redux-utils
dependency that the slice imports createThunk from, and prune the
matching tsconfig references.

juriczech committed:
chore(suite-native): simplify biometrics slice

juriczech committed:
chore(suite-native): add unit tests

peter-sanderson commented on suite-native/biometrics/src/components/BiometricsModalRenderer.tsx:
```diff
@@ -1,15 +1,48 @@
+    const shouldUserBeAuthenticated = useSelector(selectShouldUserBeAuthenticated);
+    const isBiometricsOptionEnabled = useSelector(selectIsBiometricsEnabled);
+
+    const handleBiometricsAppStateChange = useCallback(
```
nit: i would not bother with `useCallback` and simply put it into `useEffect` directly. 

This is the sole purpose of that `useEffect` hook ir map `addEventListener` subscription onto business logic. I see no point of spliting it

juriczech replied:
It's used twice in the `useEffect`, so having it as a callback makes sense and as a rule-of-hooks & no-use-before-define, it needs to be `useCallback`

Lemonexe replied:
@juriczech true. But I think Peter meant: define the function locally **within** the useEffect?
Current code is correct, but potentially dangerous: if you would change `handleBiometricsAppStateChange` – introduce a new dependency, it would cause useEffect to recalculate, not noticeable at first glance.
Inside useEffect this would of course be the same, but more easily noticeable that "oh, I need to add a new dependency to useEffect? Then I need to do things differently."

juriczech replied:
But also, then the useEffect becomes more expensive to rerun. 

However these are all microoptimizations so since you both seem to prefer to place it in the useEffect, I will do that...

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

peter-sanderson commented on suite-native/storage/src/tests/migrations/biometricsV1.test.ts:
```diff
@@ -0,0 +1,100 @@
+        ...overrides,
+    });
+
+    it('should migrate true value from storage to Redux state', () => {
```
nit: I think we do ommit should 
```suggestion
    it('migrates true value from storage to Redux state', () => {
```

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+import { getIsBiometricsFeatureAvailable, getShouldRevokeAuth } from './biometricsUtils';
+import { AuthenticateError } from './types';
+
+export enum BiometricsToggleResult {
+    Enabled = 'enabled',
+    Disabled = 'disabled',
+    BiometricsNotAvailable = 'biometrics-not-available',
+}
```
Tiny nit, just personal preference of string unions over enum.
Feel free to close this:
```suggestion
export type BiometricsToggleResult = 'enabled' | 'disabled' | 'biometrics-not-available';
```

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+    const isBiometricsAvailable = await getIsBiometricsFeatureAvailable();
+
+    if (!isBiometricsAvailable) {
+        return rejectWithValue(AuthenticateError.BiometricsNotAvailable);
```
much better than previous code, where `authenticate` simply returned nothing 👍

Lemonexe commented on suite-native/biometrics/src/biometricsSlice.ts:
```diff
@@ -0,0 +1,73 @@
+            })
+            .addCase(authenticateUserThunk.rejected, (state, { payload }) => {
+                if (payload === AuthenticateError.AuthenticationFailed) {
+                    state.biometricsError = payload ?? null;
```
can't be nullish inside the condition
```suggestion
                    state.biometricsError = payload;
```

juriczech replied:
Nice simplification, thanks!
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/biometricsSlice.ts:
```diff
@@ -0,0 +1,73 @@
+
+                if (payload === BiometricsToggleResult.Disabled) {
+                    state.isBiometricsEnabled = false;
+                }
```
I'd add comment `// BiometricsToggleResult.BiometricsNotAvailable is not considered a state change, but an unsuccessful execution, and only emits a warning alert.`
to make it clear why it is not exhaustive

juriczech replied:
After simplified return type, this was simplified even further, which would make the comment redundant IMO: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/components/BiometricOverlay.tsx:
```diff
@@ -25,28 +28,34 @@ const bottomWrapperStyle = prepareNativeStyle(utils => ({
+    const shouldUserBeAuthenticated = useSelector(selectShouldUserBeAuthenticated);
+
+    const triggerManualAuthentication = () => {
+        if (shouldUserBeAuthenticated) {
```
`shouldUserBeAuthenticated` is always true, because `BiometricOverlay` is only rendered when true.
And if it were false, we have broken UI (button doesn't do anything).
In that case I'd omit the selector completely here, it only adds confusion.
Let's make it an assumption here in `BiometricOverlay` that it is only rendered when true.

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+export const toggleBiometricsSettingsThunk = createThunk<
+    BiometricsToggleResult,
+    void,
+    { rejectValue: BiometricsToggleResult | AuthenticateError }
```
Nit: another case for string union.
In reject value you only need `BiometricsNotAvailable`, but enum does not easily let you extract a portion of it. So now it is typed that rejection can be also of type `Enabled` or `Disabled`, which can't happen.

If `BiometricsToggleResult` and `AuthenticateError` were string unions, you could type
`rejectValue: 'biometrics-not-available' | AuthenticateError`.

And vice-versa: fullfill cannot be of type `BiometricsToggleResult` but type interface makes it seems so.
That's what I was confused about in [my comment in slice](https://github.com/trezor/trezor-suite/pull/23224/changes#r3227788177), now I see it is meant that way, but if it was precisely typed, the slice matcher would indeed be exhaustive = clear.

juriczech replied:
We can use `Extract<>` on enums too. If this is good enough for you, I'd keep it. If not, let's go for string union.
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe replied:
Perfect. Enums are much smarter than I thought :clap:

Lemonexe commented on suite-native/biometrics/src/components/BiometricsModalRenderer.tsx:
```diff
@@ -1,15 +1,48 @@
+        return () => {
+            subscription.remove();
+        };
+    }, [dispatch, handleBiometricsAppStateChange, isBiometricsOptionEnabled]);
```
`dispatch` is redundant dependency.. But if you move the callback inside here, it won't be 😄 

Also, TIL that eslint `react-hooks/exhaustive-deps` does not mind unused dependencies??

juriczech replied:
That's by design I think, but very unfortunate design IMO

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+
+export const handleBiometricsAppStateChangeThunk = createThunk<
+    BiometricsAppStateChangePayload | undefined,
+    { currentAppState: AppStateStatus },
```
this is rather `nextAppState`, isn't it?

juriczech replied:
Speculative, it's `currentAppState` and `previousAppState` IMO 😄 , but let's stick to RN naming, which uses `nextAppState`. I'll rename it.

Lemonexe replied:
it is the newly changed appState so I don't think it's previous.. But I'd like `nextAppState` mainly for the consistency with other interfaces :slightly_smiling_face:

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+export const handleBiometricsAppStateChangeThunk = createThunk<
+    BiometricsAppStateChangePayload | undefined,
+    { currentAppState: AppStateStatus },
+    { rejectValue: BiometricsAppStateChangeRejectReason }
```
note that the `rejectValue` is not handled, it's just dropped. It wasn't handled before either.

juriczech replied:
Yes, I kept previous behavior. Would be nice to handle this better, but probably keep it as a follow up, if even that.

Lemonexe replied:
oki then :+1:

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+
+        switch (currentAppState) {
+            case 'active':
+                if (goneToBackgroundAtTimestamp !== null && !shouldRevokeAuth) {
```
Can't you change `shouldRevokeAuth` to return `true` when `goneToBackgroundAtTimestamp` is `null`? 
It's not used anywhere else.
Then you can simplify this condition:
```suggestion
                if (!shouldRevokeAuth) {
```

juriczech replied:
Great point. That simplifies things nicely!
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+
+            default:
+                return rejectWithValue('unhandled-app-state');
+        }
```
about the whole switch: I'm quite confused, because this is not 1:1 implementation.
There are some changes:
1. when changing to active and not within timelimit, previously it only invalidated authentication if you went from `background` to `active`, now also from `inactive` to `active`. I don't really know what it means.
2. `goneToBackgroundAtTimestamp` was only set when going to background, now also going `inactive`

Both changes SGTM, it makes more sense then before, but I don't understand details and  wonder if they are intended.

juriczech replied:
Yeah, all intended. Previous version was super unintuitive and a hell to read through IMHO. I tried to simplify it and with your comments I believe it's even better now. 

The absent 1:1 in this case is one of the main reasons for this refactor, but of course, worth mentioning and discussing.

Lemonexe commented on suite-native/biometrics/src/biometricsThunks.ts:
```diff
@@ -0,0 +1,150 @@
+            case 'inactive':
+                if (isTogglingBiometricsInProgress) {
+                    return { goneToBackgroundAtTimestamp: Date.now() };
+                }
```
I don't understand why this condition at all.
If it goes inactive during toggling biometrics, I think this thunk shouldn't care, and set `isUserAuthenticated: false`. If toggling finishes, then biometrics is turned off and it doesn't matter.
If toggling fails, then we are left authenticated when we shouldn't.
Maybe what I'm describing is an ultra edge case, but I'm trying to understand why is this code written 🙂

Lemonexe replied:
explained during call, pls add comemnt :pray:

juriczech replied:
Comment added.
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe commented on suite-native/biometrics/src/__tests__/biometricsThunks.test.ts:
```diff
@@ -0,0 +1,249 @@
+                        enabled: expectedIsBiometricsEnabled,
+                        origin: 'settingsToggle',
+                    },
+                });
```
I would add
`expect(LocalAuthentication.authenticateAsync).toHaveBeenCalled();`
I think it's an important specification that toggling is only possible if it prompts you for the biometrics check.

juriczech replied:
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

Lemonexe reviewed (commented):
Mostly LGTM, but I'd need a sync to understand some of the changes.

Lemonexe reviewed (approved):
Looks great now!
